### ISSUE TITLE
[Automation] Is the CLI `cwd` aware?

### ISSUE THREAD

**Issue by tumido** [operate-first/SRE#309]
I actually think we may include the updates to `argocd-app` as part of this (and thus the automate the full onboarding process), by just requiring the invoker to provide a path to the argocd-apps cloned repo. wdyt?

_Originally posted by @HumairAK in https://github.com/operate-first/SRE/issues/305#issuecomment-840108267_

That depends if we want the cli to be context aware - like having a config file with paths to `apps` and `argocd-apps` and so on. Or do we want the cli to be "stateless" and oblivious to its `cwd`? I think I like the idea of having a shared config file and use the cli from any place in the system, that sounds nice to me. However this highers the entry cost - you need to set up the cli before you can use it. 

I think we can do a wizard on `cli init` or something asking for the paths..

**Comment by HumairAK** [comment]
Well, it's not unheard of to have something like this, `GOPATH` comes to mind. I think I'm for context awareness. Along with `argocd-apps` we could also benefit by adding some commands to make changes to the website repo. i.e. adding markdowns/pages from other repos.

**Comment by tumido** [comment]
I agree with that. Let's design it extendable from the beginning. I'd like it to also offer a simple `cd` functionality that it would be able to change directory to those repos from anywhere. I think that would be neat.

Let's give this issue a few moments to settle and get the opinion from @larsks and @4n4nd as well.

**Comment by HumairAK** [comment]
Agreed -- I would encourage a response from others [here](https://github.com/operate-first/SRE/issues/299) as well, since this is all contingent upon going the cli route.

**Comment by larsks** [comment]
> ? I think I like the idea of having a shared config file and use the cli from any place in the system, that sounds nice to me.

I think that's an awful idea: I've been bitten so many times by things that have a single central configuration, when I want to have different configurations by directory/project/etc.

I'd prefer something that is *repository* aware, so that you can run it from anywhere within a working tree and it will find the top-level repository configuration file.

> I'd like it to also offer a simple cd functionality that it would be able to change directory to those repos from anywhere. 

I'm not entirely clear what you're suggesting here. I mean, think we should be able to point the tooling at a specific directory if we don't want it to use the cwd as the context, but that's not really a "cd functionality".

**Comment by HumairAK** [comment]
> I'd prefer something that is repository aware, so that you can run it from anywhere within a working tree and it will find the top-level repository configuration file.

@larsks can you expand on this, it sounds interesting but I'm not sure I fully understand how this would work, are you talking about something like: 

```
operate-first-org/
  - apps/
  - argocd-apps/
  - repoconfig.yaml
```

**Comment by larsks** [comment]
@HumairAK something like that, yes. And if you run the tooling from inside a subdirectory (like `apps`), it will find the top-level `repoconfig.yaml`. Generally, something that behaves like `git` with respect to configuration.

**Comment by HumairAK** [comment]
Yeah I personally like that better.

**Comment by tumido** [comment]
Hm.. I'm thinking about a tool that works rather like `kubectl` than `git`. That there's a single `.kube/config` with a list of contexts where you can choose from which context you're working with.

I think I want the CLI to be able to perform an imaginary  `add-app` command by adding manifests to the `apps` repo and create an ArgoCD app for it in the same step.

Generally speaking `git` behaves  the same no matter which folder you're at. If it's a git repo, you can do stuff in there and the result is always the same - creating a branch, sure. It creates a branch in the current repo.

On the other hand we want it here to behave differently, depending on which repo it is located at. A command can apply for a certain repo but is totally unacceptable to be executed against another repo. Can you execute "update argocd command" in the "apps" repo? Or can you run "add user" in the argocd apps? No, you can't because the CLI is picky if it will work or not, based on the repo from which you execute it. In that sense I'd like it to be able to locate the appropriate repository itself and do those changes.

I think we can make this `cwd` aware cli to work even in the scenarios like you suggest, @larsks. Simply traverse the current folder back to root unless a config file is found.  If there's none, use a default one in `~/.config` or something. This way you can have multiple config files (one in the `operate-first-org/` another in the `lars-org/` for example).

**Comment by tumido** [comment]
I also don't like to bloat each supported repo with yet another config file that we have to maintain and do backwards compatible changes to its structure. Doing this across all the repos, for different versions of the cli tool and what not seems like a big overhead. I don't think this "config" should be part of VCS and should be personal to the user's folder structure and habits.

**Comment by larsks** [comment]
What sort of configuration are we talking about?  If it *is* project specific, I don't want to have to:

```
cd /path/to/some_project
opf-cli switch-context some-project
```

I'd rather have the appropriate configuration take effect because I am operating in the specific project directory.

But honestly, I think this question is premature: we don't have a unified cli yet, nor do we have a description of what sort of configuration options we want to support. I would probably just close this issue -- or at least punt it into the future -- until after we have something more concrete to discuss.

**Comment by tumido** [comment]
I wouldn't stop it now when it seems we're converging to a resolution :stuck_out_tongue: 

I think I may have misunderstood some of the comments above and that causes further confusion. hehe, sorry :smile: 

I think this is reasonable:
```
operate-first-org/
  - apps/
  - argocd-apps/
 
  - repoconfig.yaml
```

Having a single config per project that takes effect once you're in any sub directories (repos) within this project directory is precisely what I've tried to promote here.

I would just be against to have those configs per repo and vcs tracked or what not - I'd like to leave that to the user discretion.

I think discussion diverted to the original topic a bit - and the original topic is crucial for designing this from the start:

The question at the beginning was, as I kinda understand it: *Do we want the cli behave differently based on which path you're executing it from?*

Example:
Command `A` creates a file at `folder/subfolder/x.yaml`.

When this command `A` is invoked from `apps` repo root do we expect the same result as if it would be executed from `apps/folder2`? -> resulting in always creating `apps/folder/subfolder/x.yaml`

Or do we want to always get results relative to the current directory? -> resulting in creating `apps/folder/subfolder/x.yaml` in the first case and `apps/folder2/folder/subfolder/x.yaml` in the second case.

Do we want the command `A` always work with given repo? So if command `A` is executed from the `argocd-apps` repo root while it's expected to work against `apps` file structure, do we expect it to fail or `cd` to the `apps` repo  and do changes there?

**Comment by HumairAK** [comment]
> Do we want the command A always work with given repo? So if command A is executed from the argocd-apps repo root while it's expected to work against apps file structure, do we expect it to fail or cd to the apps repo and do changes there?

I don't know if we'd actually be `cd`'ing into anyting, instead executing actions against paths relative to some location (probably where the config file is). But yeah, I think the action should yield the same result regardless of what repo you are in. Which to me makes most sense, since most of the automation work we have in the backlog are made against specific repos only and wouldn't make sense otherwise. I also don't foresee having very many actions that are similar in nature in multiple repos (save for adding docs, but we could work around this), since their purpose is often very different.

**Comment by tumido** [comment]
> I don't know if we'd actually be cd'ing into anyting

Yup.. `cd` was just a shortcut for me. :slightly_smiling_face:  it may be like `pushd/popd` or something.. That's an implementation detail to me rn.
